src/Cmd/Add.cc

```diff
@@ -0,0 +1,206 @@
+) {
+  ++itr;
+  if (itr == end) {
+    return Subcmd::missingArgumentForOpt(*--itr);
+  }
+  arg = String(*itr);
+  return None;
+}
+
+static void
+handleDependency(HashSet<StringRef>& newDeps, const StringRef dep) {
+  if (newDeps.find(dep) != newDeps.end()) {
+    logger::warn("The dependency `", dep, "` is already in the poac.toml");
+  } else {
+    newDeps.insert(dep);
+  }
+}
+
+static String
+getDependencyGitUrl(const StringRef dep) {
+  if (dep.find("://") == String::npos) {
+    // check if atleast in "user/repo" format
+    if (dep.find('/') == String::npos) {
+      logger::error("Invalid dependency: " + String(dep));
+      return "";
+    }
+
+    return "https://github.com/" + String(dep) + ".git";
+  }
+  return String(dep);
+}
+
+static String
+getDependencyName(const StringRef dep) {
+  String name;
+  if (dep.find("://") == String::npos) {
+    name = dep.substr(dep.find_last_of('/') + 1);
+  } else {
+    name = dep.substr(
+        dep.find_last_of('/') + 1, dep.find(".git") - dep.find_last_of('/') - 1
+    );
+  }
+
+  // Remove trailing '.git' if it exists
+  if (name.ends_with(".git")) {
+    name = name.substr(0, name.size() - 4);
+  }
+
+  return name;
+}
+
+static int
+addDependencyToManifest(
+    const HashSet<StringRef>& newDeps, bool isSystemDependency, String& version,
+    String& tag, String& rev, String& branch
+) {
+  toml::value depData = toml::table{};
+
+  if (isSystemDependency) {
+    if (version.empty()) {
+      logger::error("The `--version` option is required for system dependencies"
+      );
+      return EXIT_FAILURE;
+    }
+    depData["version"] = version;
+    depData["system"] = true;
+  } else {
+    if (!tag.empty()) {
+      depData["tag"] = tag;
+    }
+    if (!rev.empty()) {
+      depData["rev"] = rev;
+    }
+    if (!branch.empty()) {
+      depData["branch"] = branch;
+    }
+  }
+
+  auto data = toml::parse(getManifestPath());
+  auto& deps = toml::find<toml::table>(data, "dependencies");
+
+  for (const auto& dep : newDeps) {
+
+    if (!isSystemDependency) {
+      const String gitUrl = getDependencyGitUrl(dep);
+      const String depName = getDependencyName(dep);
+
+      if (gitUrl.empty() || depName.empty()) {
+        return EXIT_FAILURE;
+      }
+
+      deps[depName] = depData;
+      deps[depName]["git"] = gitUrl;
+    } else {
+      deps[String(dep)] = depData;
+    }
+  }
+
+  std::ofstream ofs(getManifestPath());
+  ofs << toml::format(data);
+
+  logger::info("Added", "to the poac.toml");
+
+  return EXIT_SUCCESS;
+}
+
+static int
+addMain(const std::span<const StringRef> args) {
+  if (args.empty()) {
+    logger::error("No dependencies to add");
+    return EXIT_FAILURE;
+  }
+
+  HashSet<StringRef> newDeps = {};
+
+  bool isSystemDependency = false;
+  String version; // Only used with system-dependencies
+
+  String tag;
+  String rev;
+  String branch;
+
+  std::map<
+      String, std::function<std::optional<
```

Comment:
std::map::find takes O(logN). We can instead use std::unordered_map, which takes amortized O(1).

```suggestion
  HashMap<
      StringRef, Fn<Option<
```